scheduler/system_util.go

```diff
@@ -0,0 +1,348 @@
+package scheduler
+
+// The structs and helpers in this file are split out of scheduler_system.go and
+// shared by the system and sysbatch scheduler. No code in the generic scheduler
+// or reconciler should use anything here! If you need something here for
+// service/batch jobs, double-check it's safe to use for all scheduler types
+// before moving it into util.go
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// materializeSystemTaskGroups is used to materialize all the task groups
+// a system or sysbatch job requires.
+func materializeSystemTaskGroups(job *structs.Job) map[string]*structs.TaskGroup {
+	out := make(map[string]*structs.TaskGroup)
+	if job.Stopped() {
+		return out
+	}
+
+	for _, tg := range job.TaskGroups {
+		for i := 0; i < tg.Count; i++ {
+			name := fmt.Sprintf("%s.%s[%d]", job.Name, tg.Name, i)
+			out[name] = tg
+		}
+	}
```

Comment:
Note to reviewers if this jumps out at you: this code is unchanged from the function originally in `util.go` and this is what inspired the spelunking that resulted in this PR. This is likely a bug because `tg.Count` is always 0 for system and sysbatch jobs, but I don't want to try to fix a bug and do a large refactoring in the same changeset. I'm working on that bug next.